### PR TITLE
supports serialization of custom backend metadata.

### PR DESCRIPTION
**Pull request by fakeYan**
Fixes #ISSUE_NUMBER

Add the serialization logic of backend metadata to the serialization of tensor, which is implemented through custom registration functions.

In #97429 , the structure backendMeta is provided in TensorImpl, and we think that this part of information may also need to be serialized for custom. 

cc @mruberry

### CODE REVIEW

File: torch/csrc/jit/serialization/pickler.h [hunk 1]
```diff
@@ -296,8 +297,17 @@ uint64_t getStorageKey(const at::Tensor& tensor);
 // otherwise return false
 bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls);
 
-// Return a map of Tensor Metadata for serialization.
-// For now, it only takes care of `conj` and `neg` bit.
+// Declare BackendMate serialization and deserialization function pointer types.
```
Comment (ezyang) [inline torch/csrc/jit/serialization/pickler.h:300]: nit: BackendMeta

Comment (fakeYan) [reply torch/csrc/jit/serialization/pickler.h]: Thanks for the correction, I have fixed such typos.

**Comment by ezyang** [comment]
I'm a bit at a loss about what to do with this, as serialization changes are very risky but we have no more engineers who can review it

**Comment by fakeYan** [comment]
> I'm a bit at a loss about what to do with this, as serialization changes are very risky but we have no more engineers who can review it

In my understanding, this branch will not be triggered in most scenarios. and the serialization correctness of this part of data needs to be guaranteed by custom who registered the method.

At the same time, it has no effect on the cross-platform nature of deserialization. In the PyTorch framework side, we only parse the parameters we know to restore

**Comment by fakeYan** [comment]
> I'm a bit at a loss about what to do with this, as serialization changes are very risky but we have no more engineers who can review it.

I think this is a good serialization extension with a high degree of freedom. There is no breaking of the original storage serialization process. And the correct serialization of extensions of backendmate information is guaranteed by the custom who needs it.

The most important thing is that the cross-backend nature of the data is not broken. When deserializing on the PyTorch framework side, only identifiable keyword information is processed.

**Comment by fakeYan** [comment]
I don't think this part of the serialized extension is that risky, would you consider merge this extension change.

**Review by albanD — COMMENTED** [review]
Note that we don't have any backward compatibility test for this AFAIK
So this would need to be looked at very carefully.

Also this BackendMeta was just added, is not serialized and I'm not sure it is actually used? Could you clarify why you need this?

**Comment by fakeYan** [comment]
> Note that we don't have any backward compatibility test for this AFAIK So this would need to be looked at very carefully.

I'm trying to provide a suitable test case for this extension, which should be completed within April 19.

> Also this BackendMeta was just added, is not serialized and I'm not sure it is actually used? Could you clarify why you need this?

I understand that all device-specific information has a meaning. When serializing, users will store and load this part of information as completely and correctly as possible. At the same time, we don't know the data structure that needs to be serialized, so this part needs to be handed over to the customer who needs it

Currently, this part of the information is stored in two places:
1. In TensorImpl，like #97429, we added backendMate information to TensorImpl.
2. In StorageImpl, like #98410, we recommend that users inherit storageimpl to implement their own data storage, and obviously backend-related information will also exist in subclasses.
Both scenarios are supported in this expansion.

For us, our data has different data formats in storageImpl (unlike last_channel format), and some of our calculations (partial convolution) may only support a specific data format, or be very fast in that format. If this part of the information is not saved, it may be that during the checkpoint training, the operator may frequently transdata to change the data format.

**Comment by ezyang** [comment]
Maybe this will help unblock. Naively, I imagine that if you have a custom tensor type, you probably want to serialize it in some special way by completely taking over the serialization process. What utility do you get out of reusing the existing TorchScript serialization code that is being modified here?

**Comment by fakeYan** [comment]
> Maybe this will help unblock. Naively, I imagine that if you have a custom tensor type, you probably want to serialize it in some special way by completely taking over the serialization process. What utility do you get out of reusing the existing TorchScript serialization code that is being modified here?

In fact, I think the current PyTorch serialization does a very good job, tensor and storage mapping, metadata and data organization. I think it's certainly great to be able to extend this part of serialization.

In this test case supplement, the changes are relatively large：

1. The test file` test/test_cpp_extensions_open_device_registration.py` has been refactored to avoid repeated registered verification, and the order of tests has been clarified, please focus mainly on the `test_open_device_serialization()` function
2. In actual tests, the influence of the inline keyword was found, resulting in the incorrect acquisition of fptr_map updated function pointers, so the functions originally implemented in the .h file were put into the .cpp file
3. In file `test/cpp_extensions/open_registration_extension.cpp`，`for_serialization` and `for_deserialization` are the example function.

**Comment by fakeYan** [comment]
Snipaste_2023-04-21_17-12-31 (2)

File: torch/csrc/jit/serialization/pickler.cpp [hunk 2]
```diff
@@ -801,4 +801,92 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
...
+
+void setTensorMetadata(
+    const at::Tensor& t,
+    c10::Dict<c10::IValue, c10::IValue> metadata_idict) {
+  std::unordered_map<std::string, bool> metadata;
+  for (auto& pair : metadata_idict) {
+    auto key = *pair.key().toString();
+    metadata[key] = pair.value().toBool();
+  }
+  setTensorMetadata(t, std::move(metadata));
```
Comment (ezyang) [inline torch/csrc/jit/serialization/pickler.cpp:889]: no change?

**Comment by ezyang** [comment]
OK, having understood the approach, I am more positive about this. Please stop using a string hash map with get/set fields though.

**Comment by ppiskorski** [comment]
Author of the BackendMeta patch here.
I think this patch only allows a concrete backend metadata implementation to contain boolean fields, which would not be sufficient in our case. Secondly at minimum the concrete backend type should be dependent on the device. I mean that getter/setted registry could be a mapping from device type to a pair of functions. Instead of hard-coding a check for `PrivateUse1` device it would be more generic to check if the registry contains backend metadata serializer/deserializer for a specific device type. For instance we are using the `HPU` key.

**Comment by fakeYan** [comment]
> I think this patch only allows a concrete backend metadata implementation to contain boolean fields, which would not be sufficient in our case. 
may be we can make metadata param from u`nordered_map<std::string, bool>` to `unordered_map<std::string, int>`，I think such extended expression capacity must be sufficient

> Secondly at minimum the concrete backend type should be dependent on the device. I mean that getter/setted registry could be a mapping from device type to a pair of functions. Instead of hard-coding a check for `PrivateUse1` device it would be more generic to check if the registry contains backend metadata serializer/deserializer for a specific device type. For instance we are using the `HPU` key.

I think this part of the extension is very easy to solve, of course, we can not write a fixed backend, we can judge by parsing, or we can add conditions.

File: torch/csrc/jit/serialization/pickler.cpp (hunk 2 above)
Comment (fakeYan) [reply torch/csrc/jit/serialization/pickler.cpp:889]: This function has overloading, I made a change in the function that will eventually be called
`void setTensorMetadata(const at::Tensor& t, std::unordered_map<std::string, bool> metadata) `， at line 857

File: torch/csrc/jit/serialization/pickler.cpp [hunk 3]
```diff
@@ -801,4 +801,92 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
   return true;
 }
 
+// Declare BackendMeta serialization and deserialization function pointer types.
+using BackendMetaPtr =
+    void (*)(const at::Tensor&, std::unordered_map<std::string, bool>&);
+
+// map to save function pointer for BackendMeta serialization.
+static std::unordered_map<std::string, void*> fptr_map;
```
[the remark this answers is not in the document]
Comment (fakeYan) [reply torch/csrc/jit/serialization/pickler.cpp:809]: Considering ppiskorski's opinion and extensibility, this part does require more logic.

**Comment by fakeYan** [comment]
> OK, having understood the approach, I am more positive about this. Please stop using a string hash map with get/set fields though.

I will update the code as soon as possible, thank you for your valuable suggestions

**Comment by fakeYan** [comment]
update the file
